### PR TITLE
Add chunking/iterator() support to PageQuerySet.specific()

### PR DESCRIPTION
**Pull request by ababic**
This small change solves one of the key limitations of `PageQuerySet.specific()`: That, unless the queryset is split beforehand using a paginator/slicing, Wagtail will attempt to gather all matching pages into memory at the same time before returning anything - whether you use `iterator()` or not. This causes some big problems on larger sites (100,000+ pages), where using something like the `contrib.sitemaps` views can use up significant resources, sometimes even killing a production app without horizontal scaling capabilities.

With these changes, you can use `iterator()` with `specific()` like you would any other queryset, and expect those same protections against memory overload. The same has quietly been available with `specific(defer=True)` for a few versions now, so this improves consistency too.

### CODE REVIEW

File: wagtail/query.py [hunk 1]
```diff
@@ -489,88 +514,144 @@ def specific_iterator(qs, defer=False):
...
+            if chunk_stop and slice_top > chunk_stop:
+                slice_top = chunk_stop
+            # Fetch one more than needed to determine if further chunks
+            # are available, without an extra database query
+            slice_top += 1
+            base_qs = qs[slice_bottom:slice_top]
+        else:
+            base_qs = qs
+
+        values = base_qs.values("pk", "content_type", *annotation_aliases)
```
Comment (ababic) [inline wagtail/query.py:533]: These lines are new or have changed

File: wagtail/query.py [hunk 2]
```diff
@@ -489,88 +514,144 @@ def specific_iterator(qs, defer=False):
...
+                    k: v for k, v in data.items() if k in annotation_aliases
+                }
+
+        pks_and_types = [[v["pk"], v["content_type"]] for v in values]
+
+        if chunk_size:
+            further_chunks_available = len(pks_and_types) > chunk_size
+            if further_chunks_available:
+                # Remove the 'over-fetched' item
+                pks_and_types.pop()
```
Comment (ababic) [inline wagtail/query.py:548]: These lines are new

File: wagtail/query.py [hunk 3]
```diff
@@ -489,88 +514,144 @@ def specific_iterator(qs, defer=False):
...
+                for annotation, value in annotations_by_pk.get(page.pk, {}).items():
+                    setattr(page, annotation, value)
+            yield page
+
+        # Exit outer loop if all results were returned
+        if not chunk_size or not further_chunks_available:
+            break
+
+        # Continue to next chunk
+        chunk_number += 1
```
Comment (ababic) [inline wagtail/query.py:620]: These lines are new

File: wagtail/query.py [hunk 4]
```diff
@@ -489,88 +514,144 @@ def specific_iterator(qs, defer=False):
...
-    else:
-        generic_pages = {}
-
-    # Yield all pages in the order they occurred in the original query.
-    for pk, content_type in pks_and_types:
-        try:
-            page = pages_by_type[content_type][pk]
-        except KeyError:
-            page = generic_pages[pk]
+    while True:
```
Comment (ababic) [inline wagtail/query.py:520]: All previous code is indented under this new loop

File: wagtail/query.py [hunk 5]
```diff
@@ -489,88 +514,144 @@ def specific_iterator(qs, defer=False):
...
+            slice_bottom = (chunk_start or 0) + ((chunk_number - 1) * chunk_size)
+            slice_top = (chunk_start or 0) + ((chunk_number * chunk_size))
+            if chunk_stop and slice_top > chunk_stop:
+                slice_top = chunk_stop
+            # Fetch one more than needed to determine if further chunks
+            # are available, without an extra database query
+            slice_top += 1
+            base_qs = qs[slice_bottom:slice_top]
+        else:
+            base_qs = qs
```
Comment (ababic) [inline wagtail/query.py:531]: These lines are new

File: wagtail/query.py (hunk 1 above)
Comment (ababic) [inline wagtail/query.py:533]: This line has changed to use `base_qs`

File: wagtail/query.py (hunk 2 above)
Comment (ababic) [inline wagtail/query.py:548]: These lines are new

Comment (ababic) [inline wagtail/query.py:548]: These lines are new

File: wagtail/query.py (hunk 5 above)
Comment (ababic) [inline wagtail/query.py:531]: These lines are new

File: wagtail/query.py (hunk 1 above)
Comment (ababic) [inline wagtail/query.py:533]: This line has changed to use `base_qs`

File: wagtail/query.py [hunk 6]
```diff
@@ -489,85 +500,131 @@ def specific_iterator(qs, defer=False):
...
+                    k: v for k, v in data.items() if k in annotation_aliases
+                }
+
+        pks_and_types = [[v["pk"], v["content_type"]] for v in values]
+
+        if chunk_size:
+            further_chunks_available = len(pks_and_types) > chunk_size
+            if further_chunks_available:
+                # Remove the 'over-fetched' item
+                pks_and_types.pop()
```
Comment (ababic) [inline wagtail/query.py:534]: The extra item is removed here, so the impact is not carried over to the specific queries

File: wagtail/query.py [hunk 7]
```diff
@@ -489,85 +500,131 @@ def specific_iterator(qs, defer=False):
...
+                for annotation, value in annotations_by_pk.get(page.pk, {}).items():
+                    setattr(page, annotation, value)
+            yield page
+
+        # Exit outer loop if all results were returned
+        if not chunk_size or not further_chunks_available:
+            break
+
+        # Continue to next chunk
+        chunk_number += 1
```
Comment (ababic) [inline wagtail/query.py:606]: These new lines are added to break out of the loop

**Comment by ababic** [comment]
@gasman I just figured out there may be something deeper in Django I can utilise here, using a cursor to ensure results originate from the same query. I think that would prevent the need to remove/replicate the slicing also. Will mark as draft again until I've worked it out.

**Comment by ababic** [comment]
@gasman okay, newer version is ready for review. Much less going on, and fed from the same `values()` queryset. Sometimes you have to build the wrong thing to know what the right thing is :)

**Review by gasman — APPROVED** [review]
Looks good to me! Always nice to be able to bring things like this into closer alignment with Django's implementation.

**Comment by kaedroho** [comment]
This looks awesome! Thanks for implementing it @ababic

**Comment by ababic** [comment]
Thank you both :)

**Pull request merged**